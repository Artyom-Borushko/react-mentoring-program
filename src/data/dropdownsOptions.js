export const movieDropdownOptions = ['edit', 'delete'];

export const moviesSortDropdownOptions = ['release date', 'title'];
