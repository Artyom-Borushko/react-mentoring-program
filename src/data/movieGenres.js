
export const listOfGenres = ['all', 'documentary', 'comedy', 'horror', 'crime'];
