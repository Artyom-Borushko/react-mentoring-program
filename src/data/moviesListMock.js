export const listOfMoviesMock = [
    { src: '/images/movies-list/pulpFiction.png', title: "Pulp Fiction", releaseYear: "2004", relevantGenres: ['action', 'crime'], rating: '9.2', duration: '2h 22min', description: 'Lorem ipsum dolor sit amet. Eum autem quod aut dolorem voluptatum nam dolores sunt. Ea velit placeat ea corporis illo rem quod nulla. Et maxime placeat ut vero placeat et harum sequi. A Quis expedita id dolores sint id repellendus atque sit suscipit obcaecati vel quia galisum quo cumque internos!', id: "uniqueMovieId1" },
    { src: '/images/movies-list/bohemianRhapsody.png', title: "Bohemian Rhapsody", releaseYear: "2003", relevantGenres: ['documentary', 'drama'], rating: '8.4', duration: '1h 27min', description: 'Lorem ipsum dolor sit amet. Eum autem quod aut dolorem voluptatum nam dolores sunt. Ea velit placeat ea corporis illo rem quod nulla. Et maxime placeat ut vero placeat et harum sequi. A Quis expedita id dolores sint id repellendus atque sit suscipit obcaecati vel quia galisum quo cumque internos!', id: "uniqueMovieId2" },
    { src: '/images/movies-list/killBillVol2.png', title: "Kill Bill: Vol 2", releaseYear: "1994", relevantGenres: ['comedy', 'crime', 'oscar'], rating: '6.2', duration: '3h 12min', description: 'Lorem ipsum dolor sit amet. Eum autem quod aut dolorem voluptatum nam dolores sunt. Ea velit placeat ea corporis illo rem quod nulla. Et maxime placeat ut vero placeat et harum sequi. A Quis expedita id dolores sint id repellendus atque sit suscipit obcaecati vel quia galisum quo cumque internos!', id: "uniqueMovieId3" },
    { src: '/images/movies-list/avengersWarOfInfinity.png', title: "Avengers: War of Infinity", releaseYear: "2004", relevantGenres: ['action', 'adventure'], rating: '8.2', duration: '0h 56min', description: 'Lorem ipsum dolor sit amet. Eum autem quod aut dolorem voluptatum nam dolores sunt. Ea velit placeat ea corporis illo rem quod nulla. Et maxime placeat ut vero placeat et harum sequi. A Quis expedita id dolores sint id repellendus atque sit suscipit obcaecati vel quia galisum quo cumque internos!', id: "uniqueMovieId4" },
    { src: '/images/movies-list/inception.png', title: "Inception", releaseYear: "2003", relevantGenres: ['action', 'adventure', 'horror'], rating: '10.0', duration: '1h 39min', description: 'Lorem ipsum dolor sit amet. Eum autem quod aut dolorem voluptatum nam dolores sunt. Ea velit placeat ea corporis illo rem quod nulla. Et maxime placeat ut vero placeat et harum sequi. A Quis expedita id dolores sint id repellendus atque sit suscipit obcaecati vel quia galisum quo cumque internos!', id: "uniqueMovieId5" },
    { src: '/images/movies-list/reservoirDogs.png', title: "Reservoir dogs", releaseYear: "1994", relevantGenres: ['crime', 'oscar'], rating: '7.7', duration: '2h 00min', description: 'Lorem ipsum dolor sit amet. Eum autem quod aut dolorem voluptatum nam dolores sunt. Ea velit placeat ea corporis illo rem quod nulla. Et maxime placeat ut vero placeat et harum sequi. A Quis expedita id dolores sint id repellendus atque sit suscipit obcaecati vel quia galisum quo cumque internos!', id: "uniqueMovieId6" },
];
