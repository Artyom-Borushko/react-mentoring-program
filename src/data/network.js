export default 'http://localhost:4000/movies';
