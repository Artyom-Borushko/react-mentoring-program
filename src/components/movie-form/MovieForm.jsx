import './movieForm.css';
import React, { useState } from 'react';
import { ErrorMessage, Form, Formik } from 'formik';
import Button from '../shared/button/Button';
import TextInput from '../shared/input/text-input/TextInput';
import Checkbox from '../shared/input/checkbox/Checkbox';
import TextArea from '../shared/input/text-area/TextArea';
import { movieGenres } from '../../data/movieGenres';
import formValidationSchema from '../../schemas/formValidationSchema';

export default function MovieForm({ onSubmit, movie }) {
  const [showDropdown, setShowDropdown] = useState(false);

  const toggleDropdown = () => {
    setShowDropdown(!showDropdown);
  };

  return (
    <Formik
      className="movie-form"
      initialValues={{
        movieTitle: (movie && movie.title) || '',
        movieReleaseDate: (movie && movie.release_date) || '',
        movieUrl: (movie && movie.poster_path) || '',
        movieRating: (movie && movie.vote_average) || '',
        movieGenres: (movie && movie.genres) || [],
        movieRuntime: (movie && movie.runtime) || '',
        movieOverview: (movie && movie.overview) || '',
      }}
      validationSchema={formValidationSchema}
      validateOnBlur={false}
      validateOnChange={false}
      onSubmit={(values, { setSubmitting }) => {
        onSubmit(values);
        setSubmitting(false);
      }}
    >
      {({ isSubmitting, resetForm }) => (
        <Form
          className="movie-form"
        >
          <div className="movie-form-content">
            <TextInput
              label="Title"
              name="movieTitle"
              type="text"
              id="movieTitle"
              placeholder="Title"
            />
            <TextInput
              label="Release date"
              name="movieReleaseDate"
              type="date"
              id="movieReleaseDate"
            />
            <TextInput
              label="Movie URL"
              name="movieUrl"
              type="text"
              id="movieUrl"
              placeholder="https://"
            />
            <TextInput
              label="Rating"
              name="movieRating"
              type="text"
              id="movieRating"
              placeholder="7.8"
            />
            <div className="input-wrapper">
              <button
                type="button"
                className="dropdown-opener"
                onClick={toggleDropdown}
              >
                Select Genres
              </button>
              {showDropdown && (
              <div className="dropdown-content">
                {movieGenres.map((genre) => (
                  <Checkbox
                    key={genre}
                    name="movieGenres"
                    id="movieGenres"
                    value={genre}
                    label={genre}
                  />
                ))}
              </div>
              )}
              <ErrorMessage name="movieGenres">
                {(msg) => <div className="form-field-validation-error">{msg}</div>}
              </ErrorMessage>
            </div>
            <TextInput
              label="Runtime"
              name="movieRuntime"
              type="text"
              id="movieRuntime"
              placeholder="minutes"
            />
            <TextArea
              label="Overview"
              name="movieOverview"
              id="movieOverview"
              placeholder="Movie description"
            />
          </div>

          <div className="movie-form-confirmation-control">
            <Button
              type="reset"
              disabled={isSubmitting}
              onClick={() => resetForm()}
            >
              Reset
            </Button>
            <Button
              type="submit"
              disabled={isSubmitting}
            >
              Submit
            </Button>
          </div>
        </Form>
      )}
    </Formik>
  );
}
