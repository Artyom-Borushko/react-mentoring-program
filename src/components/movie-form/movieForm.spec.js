import {render, screen} from "@testing-library/react";
import React from "react";
import userEvent from "@testing-library/user-event";
import '@testing-library/jest-dom';
import MovieForm from "./MovieForm";
import {listOfMoviesMock} from "../../data/moviesListMock";

describe("MovieForm component tests", () => {

    it("renders empty MovieForm component", () => {
        const {container} = render(<MovieForm/>);

        expect(container).toMatchSnapshot();
    });

    it("renders empty MovieForm component with initial movie", () => {
        const {container} = render(<MovieForm initialMovie={listOfMoviesMock[0]}/>);

        expect(container).toMatchSnapshot();
    });

    it("submits form with movie data", async () => {
        let submittedForm;
        const mockCallback = jest.fn(event => submittedForm = event);
        render(<MovieForm
            onSubmit={mockCallback}
        />);

        const titleInput = screen.getByLabelText('Title');
        await userEvent.click(titleInput);
        await userEvent.keyboard('test title');
        expect(titleInput.value).toBe('test title');


        const releaseDateInput = screen.getByLabelText('Release date');
        await userEvent.click(releaseDateInput);
        await userEvent.keyboard('2010-12-12');
        expect(releaseDateInput.value).toBe('2010-12-12');

        const movieUrlInput = screen.getByLabelText('Movie URL');
        await userEvent.click(movieUrlInput);
        await userEvent.keyboard('https://www.google.com/maps/');
        expect(movieUrlInput.value).toBe('https://www.google.com/maps/');

        const ratingInput = screen.getByLabelText('Rating');
        await userEvent.click(ratingInput);
        await userEvent.keyboard('6.2');
        expect(ratingInput.value).toBe('6.2');


        await userEvent.click(screen.getByRole('button', {name: 'Select Genres'}));
        await userEvent.click(screen.getByDisplayValue('Family'));
        await userEvent.click(screen.getByDisplayValue('Action'));

        const runtimeInput = screen.getByLabelText('Runtime');
        await userEvent.click(runtimeInput);
        await userEvent.keyboard('77');
        expect(runtimeInput.value).toBe('77');

        const descriptionInput = screen.getByLabelText('Overview');
        await userEvent.click(descriptionInput);
        await userEvent.keyboard('test movie description');
        expect(descriptionInput.value).toBe('test movie description');


        await userEvent.click(screen.getByRole('button', {name: 'Submit'}));
        expect(submittedForm).toEqual( {
            movieTitle: 'test title',
            movieReleaseDate: '2010-12-12',
            movieUrl: 'https://www.google.com/maps/',
            movieRating: '6.2',
            movieGenres: [ 'Family', 'Action' ],
            movieRuntime: '77',
            movieOverview: 'test movie description'
        });
    });
});
