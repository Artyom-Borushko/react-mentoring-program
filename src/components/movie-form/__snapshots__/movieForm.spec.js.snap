// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieForm component tests renders empty MovieForm component 1`] = `
<div>
  <form
    action="#"
    class="movie-form"
  >
    <div
      class="movie-form-content"
    >
      <div
        class="input-wrapper"
      >
        <label
          for="movieTitle"
        >
          Title
        </label>
        <input
          class="text-input"
          id="movieTitle"
          name="movieTitle"
          placeholder="Title"
          type="text"
          value=""
        />
      </div>
      <div
        class="input-wrapper"
      >
        <label
          for="movieReleaseDate"
        >
          Release date
        </label>
        <input
          class="text-input"
          id="movieReleaseDate"
          name="movieReleaseDate"
          type="date"
          value=""
        />
      </div>
      <div
        class="input-wrapper"
      >
        <label
          for="movieUrl"
        >
          Movie URL
        </label>
        <input
          class="text-input"
          id="movieUrl"
          name="movieUrl"
          placeholder="https://"
          type="text"
          value=""
        />
      </div>
      <div
        class="input-wrapper"
      >
        <label
          for="movieRating"
        >
          Rating
        </label>
        <input
          class="text-input"
          id="movieRating"
          name="movieRating"
          placeholder="7.8"
          type="text"
          value=""
        />
      </div>
      <div
        class="input-wrapper"
      >
        <button
          class="dropdown-opener"
          type="button"
        >
          Select Genres
        </button>
      </div>
      <div
        class="input-wrapper"
      >
        <label
          for="movieRuntime"
        >
          Runtime
        </label>
        <input
          class="text-input"
          id="movieRuntime"
          name="movieRuntime"
          placeholder="minutes"
          type="text"
          value=""
        />
      </div>
      <div
        class="input-wrapper textarea"
      >
        <label
          for="movieOverview"
        >
          Overview
        </label>
        <textarea
          class="text-area"
          id="movieOverview"
          name="movieOverview"
          placeholder="Movie description"
        />
      </div>
    </div>
    <div
      class="movie-form-confirmation-control"
    >
      <button
        class="button submit  "
        type="submit"
      >
        Reset
      </button>
      <button
        class="button submit  "
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</div>
`;

exports[`MovieForm component tests renders empty MovieForm component with initial movie 1`] = `
<div>
  <form
    action="#"
    class="movie-form"
  >
    <div
      class="movie-form-content"
    >
      <div
        class="input-wrapper"
      >
        <label
          for="movieTitle"
        >
          Title
        </label>
        <input
          class="text-input"
          id="movieTitle"
          name="movieTitle"
          placeholder="Title"
          type="text"
          value=""
        />
      </div>
      <div
        class="input-wrapper"
      >
        <label
          for="movieReleaseDate"
        >
          Release date
        </label>
        <input
          class="text-input"
          id="movieReleaseDate"
          name="movieReleaseDate"
          type="date"
          value=""
        />
      </div>
      <div
        class="input-wrapper"
      >
        <label
          for="movieUrl"
        >
          Movie URL
        </label>
        <input
          class="text-input"
          id="movieUrl"
          name="movieUrl"
          placeholder="https://"
          type="text"
          value=""
        />
      </div>
      <div
        class="input-wrapper"
      >
        <label
          for="movieRating"
        >
          Rating
        </label>
        <input
          class="text-input"
          id="movieRating"
          name="movieRating"
          placeholder="7.8"
          type="text"
          value=""
        />
      </div>
      <div
        class="input-wrapper"
      >
        <button
          class="dropdown-opener"
          type="button"
        >
          Select Genres
        </button>
      </div>
      <div
        class="input-wrapper"
      >
        <label
          for="movieRuntime"
        >
          Runtime
        </label>
        <input
          class="text-input"
          id="movieRuntime"
          name="movieRuntime"
          placeholder="minutes"
          type="text"
          value=""
        />
      </div>
      <div
        class="input-wrapper textarea"
      >
        <label
          for="movieOverview"
        >
          Overview
        </label>
        <textarea
          class="text-area"
          id="movieOverview"
          name="movieOverview"
          placeholder="Movie description"
        />
      </div>
    </div>
    <div
      class="movie-form-confirmation-control"
    >
      <button
        class="button submit  "
        type="submit"
      >
        Reset
      </button>
      <button
        class="button submit  "
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</div>
`;
